clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -723,55 +826,30 @@ final void fail(long now, Throwable throwable) {
                 // If the call was aborted while in flight due to a timeout, deliver a
                 // TimeoutException. In this case, we do not get any more retries - the call has
                 // failed. We increment tries anyway in order to display an accurate log message.
-                tries++;
                 failWithTimeout(now, throwable);
                 return;
             }
-            // If this is an UnsupportedVersionException that we can retry, do so. Note that a
-            // protocol downgrade will not count against the total number of retries we get for
-            // this RPC. That is why 'tries' is not incremented.
-            if ((throwable instanceof UnsupportedVersionException) &&
-                     handleUnsupportedVersionException((UnsupportedVersionException) throwable)) {
-                log.debug("{} attempting protocol downgrade and then retry.", this);
-                runnable.enqueue(this, now);
-                return;
-            }
-            tries++;
-            nextAllowedTryMs = now + retryBackoffMs;
 
             // If the call has timed out, fail.
             if (calcTimeoutMsRemainingAsInt(now, deadlineMs) < 0) {
                 failWithTimeout(now, throwable);
                 return;
             }
-            // If the exception is not retriable, fail.
-            if (!(throwable instanceof RetriableException)) {
-                if (log.isDebugEnabled()) {
-                    log.debug("{} failed with non-retriable exception after {} attempt(s)", this, tries,
-                        new Exception(prettyPrintException(throwable)));
-                }
-                handleFailure(throwable);
-                return;
-            }
-            // If we are out of retries, fail.
-            if (tries > maxRetries) {
-                failWithTimeout(now, throwable);
-                return;
-            }
+
             if (log.isDebugEnabled()) {
-                log.debug("{} failed: {}. Beginning retry #{}",
-                    this, prettyPrintException(throwable), tries);
+                log.debug("{} failed with non-retriable exception after {} attempt(s)", this,
+                        callRetryContext().tries(), new Exception(prettyPrintException(throwable)));
             }
-            runnable.enqueue(this, now);
+            handleFailure(throwable);
         }
 
         private void failWithTimeout(long now, Throwable cause) {
```

Comment:
I think we can simplify the `fail` logic even further by combining `failWithTimeout` and `fail` and converting `failWithTimeout` to a boolean that determines whether or not the call should be failed with a timeout like so:

```java
final void fail(long now, Throwable throwable) {
    if (failWithTimeout()) {
        log.debug("{} timed out at {} after {} attempt(s)", this, now,
                    this.callRetryContext.tries() + 1,
                    new Exception(prettyPrintException(throwable)));
        handleFailure(new TimeoutException(this + " timed out at " + now
                + " after " + this.callRetryContext.tries() + " attempt(s)", throwable));
    } else {
        handleFailure(throwable);
    }
}

private boolean failWithTimeout() {
    return aborted ||
    calcTimeoutMsRemainingAsInt(now, deadlineMs) < 0 ||
    this.callRetryContext.tries() > maxRetries;
}
```

And then we can replace all occurrences of `failWithTimeout` with just `fail`.